shell/common/rasterizer.cc

```diff
@@ -176,74 +189,71 @@ GrDirectContext* Rasterizer::GetGrContext() {
+  if (found == last_successful_tasks_.end()) {
+    return nullptr;
+  }
+  return found->second.layer_tree.get();
 }
 
 void Rasterizer::DrawLastLayerTree(
     std::unique_ptr<FrameTimingsRecorder> frame_timings_recorder) {
-  if (!last_layer_tree_ || !surface_) {
+  if (last_successful_tasks_.empty() || !surface_) {
     return;
   }
-  RasterStatus raster_status = DrawToSurface(
-      *frame_timings_recorder, *last_layer_tree_, last_device_pixel_ratio_);
+  std::list<LayerTreeTask> tasks;
+  for (auto& [view_id, task] : last_successful_tasks_) {
+    tasks.push_back(std::move(task));
+  }
+  last_successful_tasks_.clear();
+
+  DoDrawResult result =
+      DrawToSurfaces(*frame_timings_recorder, std::move(tasks));
 
   // EndFrame should perform cleanups for the external_view_embedder.
   if (external_view_embedder_ && external_view_embedder_->GetUsedThisFrame()) {
-    bool should_resubmit_frame = ShouldResubmitFrame(raster_status);
+    bool should_resubmit_frame = ShouldResubmitFrame(result);
     external_view_embedder_->SetUsedThisFrame(false);
     external_view_embedder_->EndFrame(should_resubmit_frame,
                                       raster_thread_merger_);
   }
 }
 
-RasterStatus Rasterizer::Draw(
-    const std::shared_ptr<LayerTreePipeline>& pipeline) {
+DrawStatus Rasterizer::Draw(const std::shared_ptr<FramePipeline>& pipeline) {
   TRACE_EVENT0("flutter", "GPURasterizer::Draw");
   if (raster_thread_merger_ &&
       !raster_thread_merger_->IsOnRasterizingThread()) {
     // we yield and let this frame be serviced on the right thread.
-    return RasterStatus::kYielded;
+    return DrawStatus::kYielded;
   }
   FML_DCHECK(delegate_.GetTaskRunners()
                  .GetRasterTaskRunner()
                  ->RunsTasksOnCurrentThread());
 
   DoDrawResult draw_result;
-  LayerTreePipeline::Consumer consumer =
-      [&draw_result, this,
-       &delegate = delegate_](std::unique_ptr<LayerTreeItem> item) {
-        // TODO(dkwingsmt): Use a proper view ID when Rasterizer supports
-        // multi-view.
-        int64_t view_id = kFlutterImplicitViewId;
-        std::unique_ptr<LayerTree> layer_tree = std::move(item->layer_tree);
-        std::unique_ptr<FrameTimingsRecorder> frame_timings_recorder =
-            std::move(item->frame_timings_recorder);
-        float device_pixel_ratio = item->device_pixel_ratio;
-        if (delegate.ShouldDiscardLayerTree(view_id, *layer_tree.get())) {
-          draw_result.raster_status = RasterStatus::kDiscarded;
-        } else {
-          draw_result = DoDraw(std::move(frame_timings_recorder),
-                               std::move(layer_tree), device_pixel_ratio);
-        }
-      };
+  FramePipeline::Consumer consumer = [&draw_result,
+                                      this](std::unique_ptr<FrameItem> item) {
+    draw_result = DoDraw(std::move(item->frame_timings_recorder),
+                         std::move(item->layer_tree_tasks));
```

Comment:
Something should probably clear out the tasks after this right?

Otherwise we have a list of tasks with nullptr...

Reply:
The tasks are moved all the way into `DrawToSurfacesUnsafe`, where some are discarded, some moved into `resubmitted_tasks`, and some stored in `last_successful_tasks_`. Since the vector is moved, there shouldn't be a list of nullptrs.

Reply:
Oh, I think I see now, this is the last spot they're used.